File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.java
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException {
        StreamTokenizer st = new StreamTokenizer(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		BST bst = new BST();

        int token = st.nextToken();

        while (token != StreamTokenizer.TT_EOF) {
            int v = (int) st.nval;

            bst.add(v);

            token = st.nextToken();
        }

        bw.append(bst.travel());
        bw.flush();
        bw.close();
    }
}

class BST {

    Node root = new Node(0);

    void add(int v) {
        Node node = root;

        while (true) {
            if (v < node.v) {
                if (node.left == null) {
                    node.left = new Node(v);
                    return;
                }

                node = node.left;
            } else {
                if (node.right == null) {
                    node.right = new Node(v);
                    return;
                }

                node = node.right;
            }
        }
    }

    String travel() {
        StringBuilder sb = new StringBuilder();

        travel(root.right, sb);

        return sb.toString();
    }

    void travel(Node node, StringBuilder sb) {
        if (node == null) return;

        travel(node.left, sb);
        travel(node.right, sb);

        sb.append(node.v);
        sb.append('\n');
    }

    private static class Node {
        int v;
        Node left = null;
        Node right = null;

        Node(int v) {
            this.v = v;
        }
    }
}

File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.kt
import java.io.StreamTokenizer

val bw = System.out.bufferedWriter()

fun main() = with(StreamTokenizer(System.`in`.bufferedReader())) {
    val nodes = ArrayList<Int>()
    var token = nextToken()

    while (token != StreamTokenizer.TT_EOF) {
        nodes += nval.toInt()
        token = nextToken()
    }

    val root = Node(nodes[0])

    for (i in 1 until nodes.size) {
        root.insert(nodes[i])
    }

    root.travel()

    bw.flush()
    bw.close()
}

class Node(val u: Int) {
    var left: Node? = null
    var right: Node? = null

    fun insert(v: Int) {
        when {
            u > v -> {
                left?.insert(v) ?: run { left = Node(v) }
            }
            else -> {
                right?.insert(v) ?: run { right = Node(v) }
            }
        }
    }

    fun travel() {
        left?.travel()

        right?.travel()

        bw.append("$u\n")
    }
}